index.rst

```diff
@@ -1,59 +1,367 @@
+The ``pipetask`` tool's ``--register-dataset-types`` option is a non-default option for exactly this reason: in a long-lived repository, re-executions of the same pipeline will eventually outnumber executions of new pipelines (especially new pipelines with new datasets), and hence ``--register-dataset-types`` should rarely be needed.
+Passing it all the time as a matter of habit is an antipattern, because it makes it easy for a typo to result in long-lived, hard-to-clean-up garbage (dataset types can be removed, but only if there are no datasets of that type).
+
+
+Provenance and Reproducibility
+------------------------------
+
+The plan for provenance in the Gen3 butler is centered around storing the directed acyclic graph (DAG) of datasets and processing "quanta" that is used to drive ``PipelineTask`` execution, after updating it with the unique identifiers of the datasets actually produced and annotating it with information about which input datasets are actually used by the (rare) ``PipelineTasks`` that may not use all predicted inputs.
+While some provenance information (e.g. software versions and configurations) are currently associated directly with ``RUN`` collections (and this information, at least, may always be), and ``CHAINED`` collections provide some information about what datasets were used as inputs when creating others (see :ref:`collections-shared-official-processing-outputs`), these do not carry sufficiently detailed information about the relationships between datasets to meet our needs.
+
+Using fine-grained provenance information to exactly reprocess a DAG will actually be quite different from starting a new run "from scratch", as it doesn't involve providing collections or data IDs as inputs - the input datasets are already fully resolved, so there is no need to search for them in collections, and the data IDs are intrinsic to those datasets.
+We will also need to provide ways to *almost* exactly reprocess a DAG, of course - e.g. replacing the initial resolved datasets with new collection + data ID searches, modifying ``Task`` configuration in a way that does not change the DAG (or changes it only in a limited sense), and probably more.
+
+All of this fine-grained provenance is not yet implemented, however, and at present the only way to guarantee reproducibility is for all input collections to have exactly the same state they had when the original run was performed.
+The standard collections defined in this document are poorly suited for this role, however; we consider it more important for these to track the "current best" (or in the case of raws, recent observations) than it is for them remain immutable.
+Users should thus be aware that repeated processing runs using the same input collections (and everything else held constant) are *not* intended to always produce the same results (and this is a feature, not a bug).
```

Comment:
+1 to everything in this provenance section: this is important, but this document only provides the foundation on which to build such provenance. I think we need to call this out loudly and often, until said reproducibility guarantees are available.